pull_request_template.md

```diff
@@ -0,0 +1,19 @@
+Hi there!
+Thanks for contributing to Swarm!
+
+Before opening a Pull Request or reviewing a Pull Request, please get yourself acquainted with the following guidelines:
+
+1. Open PRs preemptively (i.e. whenever you start working on it), especially for the case of large tasks
+2. Create PRs with a healthy amount of description. A PR without a description will not be approved
+3. PR names should adhere to the following standard: "(affected package(s)): description". examples: "storage: fix panic on hasherstore put", "network, storage, p2p: integrate capabilities api". if many (many) packages are affected, use "all" instead of individual names
+3. Mark the PR with a `ready for review` label once it is. If a second(or more) round of reviews is required, remove the `ready for review` label and add a `ready for another review` label
+4. Never(!) force-push a PR once reviews have started! This has the unfortunate side-effect of resolving/outdating previous comments over the PR
```

Comment:
so rebase only in the end?

Reply:
ideally yes. and ideally you shouldn't need to rebase at all. i'm willing to reconcile with adding "rebasing while the pr is under review is to be done only if requested by reviewers"

Reply:
this is more complicated. if the PR review is a longer process, it is super important to keep it rebased, since it can bring up incompatibilities etc.
Another reason not to keep review process too long and PRs to be small. 
@janos you usually do a merge what do u think is the best way?

Reply:
This point is fine as a warning, even I doubt that it can be always applied, but at least it would minimize force pushing. As we cannot restore old history after force pushing, I do not see what we can do when that happens. It has a bit of conflict with point number 1. With preemptive PRs, we are encouraging early comments, which means that this point would be activated very early. 

I think that rebasing and force pushing breaks history for other clones, especially when conflicts need to be resolved afterwards repeatedly. Having a merge commit allows better automatic conflict resolution by not loosing commits and history. But, I have nothing against force pushing as it provides cleaner history. It is completely ok to do force push for me, as it is a usual operation on swarm codebase. I would prefer forbidding force pushing completely, as we are squashing commits at the end, but that would be too drastic.

Reply:
Should we encourage development in parallell temporary files where possible? That way we have less issues with rebase.

Reply:
My idea is to avoid force pushing as much as possible, that is why I added the reservations that the reviewer can ask to do so once the review process has started.

Rebasing brings a plethora of problems ranging from regressions due to human error while rebasing to github quirks that resolve comments and other branches that break while rebasing. These thing happen to us all the time and I would like to reduce the number of points of possible failures. We shouldn't rebase unless another branch contains a changeset that is needed in the current one we're working on. If master changes in the meanwhile when we work on the PR - it does not infer that a rebase is in need. If our CI runs a `merge` build and not a `push` build the incompatibilities should pop up and be shown on the CI build.

I agree that this point somehow is in conflict with point 1, but the idea is to give guidelines to contributors while allowing core members a bit more flexibility according to fluctuating needs. There's no silver bullet here but we can at least minimize error by new and external contributors.

@nolash you are welcome to PR or start a document on development best practices. I really don't wanna make a PR template into a gocode best practices knowledge-base

Reply:
@acud my point is not a go-specific one. Conflicts would happen in any kind of source. By keeping separate temporary files, rebase is even less likely. This is also aligned with your announced goal:

> My idea is to avoid force pushing as much as possible